custom/zipline/views.py

```diff
@@ -0,0 +1,509 @@
+        pass
+
+
+class DeliveredStatusUpdateView(BaseZiplineStatusUpdateView):
+
+    def validate_and_clean_payload(self, order, data):
+        self.validate_and_clean_int(data, 'packageNumber')
+
+    def process_status_update(self, order, data):
+        dispatched_data = self.carry_forward_dispatched_data(order, data['packageNumber'])
+        delivered_status = EmergencyOrderStatusUpdate.create_for_order(
+            order.pk,
+            EmergencyOrderStatusUpdate.STATUS_DELIVERED,
+            zipline_timestamp=data['timestamp'],
+            package_number=data['packageNumber'],
+            **dispatched_data
+        )
+
+        update_product_quantity_json_field(order.products_delivered, dispatched_data['products'])
+
+        if not order.delivered_status and self.all_flights_done(order):
+            order.status = EmergencyOrderStatusUpdate.STATUS_DELIVERED
+            order.delivered_status = delivered_status
+
+        order.save()
+
+        package = self.get_package_object(order, data['packageNumber'])
+        if not package.delivered_status:
+            package.status = EmergencyOrderStatusUpdate.STATUS_DELIVERED
+            package.delivered_status = delivered_status
+            package.save()
+
+        return True, {'status': 'success'}
+
+    def send_sms_for_status_update(self, order, data):
+        pass
+
+
+class CancelledInFlightStatusUpdateView(BaseZiplineStatusUpdateView):
+
+    def validate_and_clean_payload(self, order, data):
+        self.validate_and_clean_int(data, 'packageNumber')
+
+    def process_status_update(self, order, data):
+        cancelled_in_flight_status = EmergencyOrderStatusUpdate.create_for_order(
+            order.pk,
+            EmergencyOrderStatusUpdate.STATUS_CANCELLED_IN_FLIGHT,
+            zipline_timestamp=data['timestamp'],
+            package_number=data['packageNumber'],
+            **self.carry_forward_dispatched_data(order, data['packageNumber'])
+        )
+
+        if not order.delivered_status and self.all_flights_done(order):
+            order.status = EmergencyOrderStatusUpdate.STATUS_DELIVERED
+            order.delivered_status = EmergencyOrderStatusUpdate.objects.filter(
+                order_id=order.pk,
+                status=EmergencyOrderStatusUpdate.STATUS_DELIVERED
+            ).order_by('-zipline_timestamp')[0]
+            order.save()
+        elif not order.cancelled_status and self.all_flights_cancelled(order):
+            order.status = EmergencyOrderStatusUpdate.STATUS_CANCELLED
+            order.cancelled_status = cancelled_in_flight_status
+            order.save()
+
+        package = self.get_package_object(order, data['packageNumber'])
+        if not package.cancelled_status:
+            package.status = EmergencyOrderStatusUpdate.STATUS_CANCELLED_IN_FLIGHT
+            package.cancelled_status = cancelled_in_flight_status
+            package.save()
+
+        return True, {'status': 'success'}
```

Comment:
you probably know about it but `[json_error](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/view_utils.py#L45)` might be useful.